kernel/src/interrupt.rs

```diff
@@ -12,20 +19,78 @@ use x86_64::{
     VirtAddr,
 };
 
+type SupportedKeyboard = Keyboard<pc_keyboard::layouts::Us104Key, pc_keyboard::ScancodeSet1>;
+
 static IDT: Once<InterruptDescriptorTable> = Once::new();
 static TSS: Once<TaskStateSegment> = Once::new();
 static GDT: Once<GlobalDescriptorTable> = Once::new();
 static SEGMENT_SELECTORS: Once<SegmentSelectors> = Once::new();
+static PICS: Once<Mutex<ChainedPics>> = Once::new();
+static KEYBOARD: Once<Mutex<SupportedKeyboard>> = Once::new();
 
 const DOUBLE_FAULT_IST_INDEX: u16 = 0;
 
+// hardware interrupts PICs (slots 32-47)
+// Safety - ensure that the PICs does not overlap
+const PIC_1_OFFSET: u8 = 32;
+const PIC_2_OFFSET: u8 = PIC_1_OFFSET + 8;
+
+const PS2_CONTROLLER_PORT: u16 = 0x60;
+
+// idk, i just think it is important to ensure safety of this unsafe block with some rust magic
+/// Wraps handler in function that calls the handler and notifies the PICs about the interrupt end
+/// then adds the handler to the IDT.
+macro_rules! interrupt_handler {
+    ($idt:tt, $interrupt_index:path, $handler:ident) => {
+        paste! {
+            extern "x86-interrupt" fn [<$handler _wrapper>] (stack: InterruptStackFrame) {
+                $handler(stack);
+
+                // # Safety
+                // we use the same interrupt number as the handler is registered for
+                unsafe {
+                    PICS.get().unwrap().lock().notify_end_of_interrupt($interrupt_index.into());
+                }
+            }
+            $idt[$interrupt_index.into()].set_handler_fn([<$handler _wrapper>]);
+        }
+    };
+}
```

Comment:
This is what I came up with: https://godbolt.org/z/EqajarooE. I'm not even sure the trait needs to be unsafe as I don't see how can it cause undefined behavior. In the worst case it will just call the wrong handler